crates/kubelet/src/state.rs

```diff
@@ -258,14 +266,62 @@ pub async fn run_to_completion<PodState: Send + Sync + 'static>(
     pod_state: &mut PodState,
     pod: Arc<RwLock<Pod>>,
 ) {
-    let (name, namespace) = {
+    let (name, api) = {
         let initial_pod = pod.read().await.clone();
         let namespace = initial_pod.namespace().to_string();
         let name = initial_pod.name().to_string();
-        (name, namespace)
+        let api: Api<KubePod> = Api::namespaced(client.clone(), &namespace);
+        (name, api)
     };
 
-    let api: Api<KubePod> = Api::namespaced(client.clone(), &namespace);
+    // NOTE: This loop patches the container statuses of the Pod with and then
+    // waits for them to be picked up by the reflector. This is needed for a
+    // few reasons:
+    // * Kubernetes rewrites an empty array to null, preventing us from
+    //   starting with that and appending.
+    // * Pod reflection is not updated within a given state, meaning that
+    //   container status patching cannot be responsible for initializing this
+    //   (this would be a race condition anyway).
+    // I'm not sure if we want to loop forever or handle some sort of failure
+    // condition (if Kubernetes refuses to accept and propagate this
+    // initialization patch.)
+    'main: loop {
```

Comment:
Ok, think I have a few concerns about this. This is mainly broken down into two things 1) I don't know if this is any better than our old method of doing things and 2) If we do continue doing it this way, I think there could be a better way of doing this. 

1. So I don't know if this provides any benefit on our previous version other than API call reduction. We still have a bunch of waiting and things that we end up doing (no matter how we implement the retry). I think it could actually end up being slower as things step through the state machine
2. If we do want to go this route, I think we should reuse/export the `retry!` macros from the node code at the very minimum. I think a more elegant solution could entail wrapping the pod in a custom struct with a `RwLock` that write locks until it receives a response from the API and then abstract that all away behind `pod.read()` like we already have

Reply:
So regarding performance, I think this is better for two reasons:
* this loop only runs when a Pod is first scheduled, not on each state change.
* in most cases, all it does is wait a second or two for the reflector to update, and we don't have to try to patch multiple times. 
I will address the other nits.

Reply:
> this loop only runs when a Pod is first scheduled, not on each state change.

Ok, so I misread that. I do think we should have a max number of retries (preferably as a `const`) like what you have now and then just error out and let k8s know the pod has errors (this might be a good use of the `Unknown` phase)

Reply:
Sounds good will do.